apis/secretsmanager/v1alpha1/secret_types.go

```diff
@@ -0,0 +1,223 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/docker/swarmkit/api/timestamp"
+)
+
+// Keys used in connection secret.
+const (
+	ConnectionSecretKeyName        = "secret"
+	ConnectionSecretKeyProjectName = "projectName"
+)
+
+// SecretVersionState gives the state of a [SecretVersion][google.cloud.secretmanager.v1.SecretVersion], indicating if it can be accessed.
+type SecretVersionState int32
+
+const (
+	// SecretVersionSTATEUNSPECIFIED represents that SecretVersion state is not specified. This value is unused and invalid.
+	SecretVersionSTATEUNSPECIFIED SecretVersionState = 0
+	// SecretVersionENABLED represents that SecretVersion state  may be accessed.
+	SecretVersionENABLED SecretVersionState = 1
+	// SecretVersionDISABLED represents that SecretVersion state may not be accessed, but the secret data
+	// is still available and can be placed back into the [ENABLED] state.
+	SecretVersionDISABLED SecretVersionState = 2
+	// SecretVersionDESTROYED represents that SecretVersion state is destroyed and the secret data is no longer
+	// stored. A version may not leave this state once entered.
+	SecretVersionDESTROYED SecretVersionState = 3
+)
+
+// SecretVersionStateName is to map integers with states
+var SecretVersionStateName = map[int32]string{
+	0: "STATE_UNSPECIFIED",
+	1: "ENABLED",
+	2: "DISABLED",
+	3: "DESTROYED",
+}
+
+// SecretVersionStateValue is to map states with integers
+var SecretVersionStateValue = map[string]int32{
+	"STATE_UNSPECIFIED": 0,
+	"ENABLED":           1,
+	"DISABLED":          2,
+	"DESTROYED":         3,
+}
+
+// SecretParameters defines parameters for a desired Secret Manager's secret.
+type SecretParameters struct {
+
+	// Required. The resource name of the project to associate with the
+	// [Secret][google.cloud.secretmanager.v1.Secret], in the format `projects/*`.
+	Parent string `json:"parent,omitempty"`
+
+	// Labels are used as additional metadata on Topic.
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// The replication policy of the secret data attached to the Secret. Can be automatic or userManaged.
+	// Ref: https://cloud.google.com/secret-manager/docs/reference/rest/v1/projects.secrets
+	// +optional
+	// +immutable
+	Replication *Replication `json:"replication,omitempty"`
+
+	// SecretVersions is a list of SecretVersion resource in the Secret Manager API.
+	SecretVersions *[]SecretVersion `json:"secretVersion,omitempty	"`
```

Comment:
`SecretVersion` should be its own CRD and have a reference to the `Secret`. This is definitely a little bit of a strange API as the call to create the `SecretVersion` actually lives under the `Secret` endpoint https://cloud.google.com/secret-manager/docs/reference/rest/v1/projects.secrets/addVersion